opt: correctly reconcile column definition list with RECORD-returning UDFs

DrewKimball opened the pull request:
#### sql: remove usages of `types.IsRecordType`

Previously, the `types.IsRecordType` function was used in different
contexts (with vs without OUT-params, function params vs return type).
This made it difficult to determine whether a particular usage was
correct, and led to a few bugs in cases where additional checks
were necessary.

This commit replaces usages of `types.IsRecordType` with either:
1. `typ.Identical(types.AnyTuple)`, or
2. `typ.Oid() == oid.T_record`

The former should be used for a RECORD-returning routine with no
OUT-parameters, as well as for a RECORD-typed variable. The latter
should be used to match either a RECORD-returning routine, or one
with multiple OUT-parameters.

Informs #114846

Release note: None

#### optbuilder: use actual arg types when building routine with wildcard types

Previously, we would always pass the static parameter types when
building the routine. However, in some cases the static type is
a wildcard, so we actually need to use the actual argument type. Note
that always using the actual argument type can be incorrect (e.g. we'd
lose the tuple labels present in the static type).

Release note: None

#### opt: refactor optbuild paths for routines and generator functions

This commit heavily refactors the type-handling logic for routines and
generator functions. The hope is to make the code more readable, and also
make the changes in the next commit easier.

Informs #114846

Release note: None

#### opt: add assignment casts for UDFs used as a data source

Previously, attempting to use a RECORD-returning UDF as a data source
(e.g. `SELECT * FROM` syntax) would result in an internal error if the
column definition list types didn't match the columns of the last
statement. This commit fixes that by adding validation that the types
are either identical or can be assignment-casted, and adding assignment
casts if necessary.

Fixes #114846
Fixes #113186

Release note (bug fix): Fixed a bug that could cause an internal error of
the form `invalid datum type given: ..., expected ...` when a RECORD-returning
UDF used as a data source was supplied a column definition list with
mismatched types. This bug has existed since v23.1.0.

#### opt/optbuilder: check for coercibility instead of tuple types

This commit changes the handling of tuple-returning routines to mirror
that of postgres. In particular, when the routine return type is a tuple,
postgres first attempts to coerce result columns to the return type of the
routine. Only if that attempt fails, postgres wraps the result column in
a tuple, and again attempts the coercion. This change affects the handling
of routines that return (for example) a single composite-typed column. For
example, the following two logic tests should produce the same result:
```
statement ok
CREATE TYPE two_typ AS (x INT, y INT);
CREATE FUNCTION f() RETURNS two_typ LANGUAGE SQL AS $$ SELECT 1, 2; $$;

query T
SELECT f();
----
(1,2)
```
vs
```
statement ok
CREATE TYPE two_typ AS (x INT, y INT);
CREATE FUNCTION f() RETURNS two_typ LANGUAGE SQL AS $$ SELECT ROW(1, 2); $$;

query T
SELECT f();
----
(1,2)
```
There is not release note, since this shouldn't affect versions prior to 24.1.

Fixes #120942

Release note: None

yuzefovich committed:
optbuilder: use actual arg types when building routine with wildcard types

Previously, we would always pass the static parameter types when
building the routine. However, in some cases the static type is
a wildcard, so we actually need to use the actual argument type. Note
that always using the actual argument type can be incorrect (e.g. we'd
lose the tuple labels present in the static type).

Release note: None

DrewKimball committed:
opt: refactor optbuild paths for routines and generator functions

This commit heavily refactors the type-handling logic for routines and
generator functions. The hope is to make the code more readable, and also
make the changes in the next commit easier.

Informs #114846

Release note: None

DrewKimball committed:
opt: add assignment casts for UDFs used as a data source

Previously, attempting to use a RECORD-returning UDF as a data source
(e.g. `SELECT * FROM` syntax) would result in an internal error if the
column definition list types didn't match the columns of the last
statement. This commit fixes that by adding validation that the types
are either identical or can be assignment-casted, and adding assignment
casts if necessary.

Fixes #114846
Fixes #113186

Release note (bug fix): Fixed a bug that could cause an internal error of
the form `invalid datum type given: ..., expected ...` when a RECORD-returning
UDF used as a data source was supplied a column definition list with
mismatched types. This bug has existed since v23.1.0.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119616)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
LGTM although I can't say I fully grasp all the details. There are some CI failures though ...

Also I'm quite sure that `pg_regress` roachtest will fail after this, so it'd be nice to update the expected diff with each commit.

> Reviewed 7 of 7 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-:-NrbOB1PC0VN3h7Q70tN:bekysag)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 740 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NrbIKguC2IYC3ORNeVa:-NrbIKguC2IYC3ORNeVb:bma538y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d028497f29e21bb06fc2c0be8344c2da87ca21/pkg/sql/catalog/funcdesc/func_desc.go#L740)):*
> ```Go
> 	}
> 	returnType := desc.ReturnType.Type
> 	if types.IsRecordReturnTypeNoOutParams(returnType) {
> ```

I'm confused about the logic here. In this case we _might_ have OUT params specified, yet we use `IsRecordReturnTypeNoOutParams` - is this because at this point `returnType` doesn't yet have labels that come from the OUT params?
___
*[`pkg/sql/opt/testutils/testcat/function.go` line 170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NrbNedI-PqRJxWp_N9S:-NrbNedI-PqRJxWp_N9T:bz0k18x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2202def42bc1b66c8521bb46ee6888cdca173ba/pkg/sql/opt/testutils/testcat/function.go#L170)):*
> ```Go
> 		Type:                  routineType,
> 	}
> 	overload.ReturnsRecordType = types.IsRecordReturnTypeNoOutParams(retType)
> ```

nit: this change belongs in the first commit.


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 1 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/types/types.go` line 2851 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NsEG-ly9nVQua5JV_ZS:-NsEG-ly9nVQua5JV_ZT:b-t2q8jz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d028497f29e21bb06fc2c0be8344c2da87ca21/pkg/sql/types/types.go#L2851)):*
> ```Go
> //
> // This should only be used when processing routines.
> func IsRecordParamType(typ *T) bool {
> ```

These two functions are the same, are they not?

Also, I don't think the higher level concepts of UDFs/routines/procedures should leak into the lower level `types` package. If we want to keep these functions here, I think we should name and describe them with type-related terminology, not routine-related terminology.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: remove usages of `types.IsRecordType`

Previously, the `types.IsRecordType` function was used in different
contexts (with vs without OUT-params, function params vs return type).
This made it difficult to determine whether a particular usage was
correct, and led to a few bugs in cases where additional checks
were necessary.

This commit replaces usages of `types.IsRecordType` with either:
1. `typ.Identical(types.AnyTuple)`, or
2. `typ.Oid() == oid.T_record`

The former should be used for a RECORD-returning routine with no
OUT-parameters, as well as for a RECORD-typed variable. The latter
should be used to match either a RECORD-returning routine, or one
with multiple OUT-parameters.

Informs #114846

Release note: None

DrewKimball commented:
I've changed the first commit to just remove usages of `IsRecordType` entirely. Now, if you want to know whether the routine _either_ returns RECORD or multiple OUT params, you check if the oid of the return type is `T_record`. If you want to know that the routine returns RECORD (with no OUT params specified), you can check that the type is identical to `types.Any` - this also works for RECORD-typed variables.
Note that once the last statement is built, the function's concrete return type is resolved, so the `types.Any` check fails. It's still possible to use `Overload.ReturnsRecordType`, though, which is now false for a routine with OUT parameters.

yuzefovich reviewed (commented):
Nice work figuring this out! I think it makes sense to me - I only have a few nits.

> Reviewed 14 of 14 files at r5, 2 of 2 files at r6, 8 of 8 files at r7, 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-:-NueXXiP1bHh_q9I4CG0:bpfm7yw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 56 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NueWoMI9xhT9UMvb_TN):*
I think it also fixes #120942. Also we have some commented out tests that reference this issue - only UDT one with procedure seems to not be yet fixed (are you planning to address that one too as part of this PR)?
___
*[`pkg/sql/opt/optbuilder/srfs.go` line 238 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NueWF1m4v2GxeWeYe5U:-NueWF1m4v2GxeWeYe5V:b-ktgkh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d14938244f1d4f1ec19a2d3489904fc0b4f38d78/pkg/sql/opt/optbuilder/srfs.go#L238)):*
> ```Go
> 		if rTyp.Oid() != oid.T_record {
> 			// Non RECORD-return type with a column definition list is not permitted.
> 			if rTyp.Family() == types.TupleFamily {
> ```

nit: should we include tests for the newly-added errors? Looks like some examples are present in the `pg_regress`.
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 792 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NueXDsDCyeTC9B4BjjE:-NueXDsDCyeTC9B4BjjF:b-k02r2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/055b9ddcaa7f2a48a46629b0bb6c744b9375c6ff/pkg/sql/logictest/testdata/logic_test/udf#L792)):*
> ```
> CREATE FUNCTION f113186() RETURNS RECORD AS $$ SELECT 1.99; $$ LANGUAGE SQL;
> 
> # Until #113186 is resolved, the internal error is expected.
> ```

nit: remove this comment.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt/optbuilder: check for coercibility instead of tuple types

This commit changes the handling of tuple-returning routines to mirror
that of postgres. In particular, when the routine return type is a tuple,
postgres first attempts to coerce result columns to the return type of the
routine. Only if that attempt fails, postgres wraps the result column in
a tuple, and again attempts the coercion. This change affects the handling
of routines that return (for example) a single composite-typed column. For
example, the following two logic tests should produce the same result:
```
statement ok
CREATE TYPE two_typ AS (x INT, y INT);
CREATE FUNCTION f() RETURNS two_typ LANGUAGE SQL AS $$ SELECT 1, 2; $$;

query T
SELECT f();
----
(1,2)
```
vs
```
statement ok
CREATE TYPE two_typ AS (x INT, y INT);
CREATE FUNCTION f() RETURNS two_typ LANGUAGE SQL AS $$ SELECT ROW(1, 2); $$;

query T
SELECT f();
----
(1,2)
```
There is not release note, since this shouldn't affect versions prior to 24.1.

Fixes #120942

Release note: None

DrewKimball commented:
I've added a commit that uncomments the test cases with TODOs and fixes the single composite return type cases. I'll add some error-tests to the third commit as well.
@yuzefovich @mgartner is this getting too unwieldy? Would you rather I split this into more than one PR?

yuzefovich reviewed (approved):
In terms of splitting up this PR - up to you and Marcus. I reviewed this PR from scratch this week, and it <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

> Reviewed 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-:-NukTnNf9FhhN4LTKw_X:b77pbtd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/optbuilder/create_function.go` line 416 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NukS9PE4xc7rYiLJ5OO:-NukS9PE4xc7rYiLJ5OP:bj0km1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6858c518a8dca737dd988d9717558c2c2a2c1e/pkg/sql/opt/optbuilder/create_function.go#L416)):*
> ```Go
> 	expected *types.T,
> 	cols []scopeColumn,
> 	isSQLProcedure bool,
> ```

nit: `isSQLProcedure` is no longer used.
___
*[`pkg/sql/logictest/testdata/logic_test/procedure` line 480 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NukTXsd6SBQZ3vDPtNx:-NukTXse4JKV1bs-xZ_v:b-s8dekk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6858c518a8dca737dd988d9717558c2c2a2c1e/pkg/sql/logictest/testdata/logic_test/procedure#L480)):*
> ```
> 
> query T
> CALL p(NULL);
> ```

nit: it seems like this CALL is redundant with the one above - perhaps remove it and move DROP PROCEDURE to be above CREATE PROCEDURE that fails?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`-- commits` line 56 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NueWoMI9xhT9UMvb_TN):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think it also fixes #120942. Also we have some commented out tests that reference this issue - only UDT one with procedure seems to not be yet fixed (are you planning to address that one too as part of this PR)?
</blockquote></details>

Done (in the last commit).
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 740 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NrbIKguC2IYC3ORNeVa:-NuyI1Y72cpSn5Hxy-Dh:b-mowpyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d028497f29e21bb06fc2c0be8344c2da87ca21/pkg/sql/catalog/funcdesc/func_desc.go#L740)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I'm confused about the logic here. In this case we _might_ have OUT params specified, yet we use `IsRecordReturnTypeNoOutParams` - is this because at this point `returnType` doesn't yet have labels that come from the OUT params?
</blockquote></details>

Done (removed `IsRecordReturnTypeNoOutParams`).
___
*[`pkg/sql/opt/optbuilder/create_function.go` line 416 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NukS9PE4xc7rYiLJ5OO:-NuyI8FVDPJZTdLANvwg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6858c518a8dca737dd988d9717558c2c2a2c1e/pkg/sql/opt/optbuilder/create_function.go#L416)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `isSQLProcedure` is no longer used.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/srfs.go` line 238 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NueWF1m4v2GxeWeYe5U:-NuyI96n1JfOob0pPmcO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d14938244f1d4f1ec19a2d3489904fc0b4f38d78/pkg/sql/opt/optbuilder/srfs.go#L238)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we include tests for the newly-added errors? Looks like some examples are present in the `pg_regress`.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/testutils/testcat/function.go` line 170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NrbNedI-PqRJxWp_N9S:-NuyIASK1QCe9xtskM4e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2202def42bc1b66c8521bb46ee6888cdca173ba/pkg/sql/opt/testutils/testcat/function.go#L170)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this change belongs in the first commit.
</blockquote></details>

Done.
___
*[`pkg/sql/types/types.go` line 2851 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NsEG-ly9nVQua5JV_ZS:-NuyII8m254ns5arobSp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d028497f29e21bb06fc2c0be8344c2da87ca21/pkg/sql/types/types.go#L2851)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

These two functions are the same, are they not?

Also, I don't think the higher level concepts of UDFs/routines/procedures should leak into the lower level `types` package. If we want to keep these functions here, I think we should name and describe them with type-related terminology, not routine-related terminology.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/procedure` line 480 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NukTXsd6SBQZ3vDPtNx:-NuyH60B5y9LGPPSDTVj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6858c518a8dca737dd988d9717558c2c2a2c1e/pkg/sql/logictest/testdata/logic_test/procedure#L480)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it seems like this CALL is redundant with the one above - perhaps remove it and move DROP PROCEDURE to be above CREATE PROCEDURE that fails?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 792 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-NueXDsDCyeTC9B4BjjE:-NuyHbHT41Bi8HfLZgJK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/055b9ddcaa7f2a48a46629b0bb6c744b9375c6ff/pkg/sql/logictest/testdata/logic_test/udf#L792)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: remove this comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice job working through these edge cases!

> Reviewed 2 of 14 files at r5, 20 of 20 files at r10, 2 of 2 files at r11, 12 of 12 files at r12, 5 of 5 files at r13, 7 of 7 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-:-Nv8wiOI6WC2kjq9814K:b-ckzpsr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/routine.go` line 366 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-Nv8vRaD0bRlE4mdcP0S:-Nv8vRaD0bRlE4mdcP0T:b7ytdeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/23a3ce07effdcb75116398e8d31706e99119b47b/pkg/sql/opt/optbuilder/routine.go#L366)):*
> ```Go
> 			argColName := funcParamColName(tree.Name(paramType.Name), i)
> 			// Use statically defined argument type (unless it is a wildcard in
> 			// which case we use the actual parameter type).
> ```

nit: I'm being really nit-picky here, but I think the language of this comment doesn't match the logic. It looks like we use the statistically defined *parameter* type (e.g., the type of the parameter in the CREATE FUNCTION statement) unless it is a wildcard type, then we use the actual *argument* type.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/udf` line 366 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-Nv8vo3V-V8PXzg2LCvP:-Nv8vo3V-V8PXzg2LCvQ:b-9jfhzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/23a3ce07effdcb75116398e8d31706e99119b47b/pkg/sql/opt/exec/execbuilder/testdata/udf#L366)):*
> ```
>   row 0, expr 0: f88259(333, 444)
> 
> # Regression test for not using actual parameter types (#114846).
> ```

nit: actual *argument* types


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119616)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/routine.go` line 366 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-Nv8vRaD0bRlE4mdcP0S:-NvGod1z3ReE5g7k9j-W:beu4u9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/23a3ce07effdcb75116398e8d31706e99119b47b/pkg/sql/opt/optbuilder/routine.go#L366)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I'm being really nit-picky here, but I think the language of this comment doesn't match the logic. It looks like we use the statistically defined *parameter* type (e.g., the type of the parameter in the CREATE FUNCTION statement) unless it is a wildcard type, then we use the actual *argument* type.
</blockquote></details>

Good point, Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/udf` line 366 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/119616#-Nv8vo3V-V8PXzg2LCvP:-NvGomlRE3oknWKD1L8X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23a3ce07effdcb75116398e8d31706e99119b47b/pkg/sql/opt/exec/execbuilder/testdata/udf#L366)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: actual *argument* types
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I just ran `pg_regress` roachtest on SHA that includes this change, and we got some new internal errors (full diff is https://github.com/yuzefovich/cockroach/commit/898ddee21be7939b2d14d174dae98825838b3f71), for example:
```sql
CREATE OR REPLACE FUNCTION dup (inout f2 anyelement, out f3 anyarray)
 AS 'select $1, array[$1,$1]' LANGUAGE sql;
 SELECT dup(22);
```

DrewKimball commented:
Hm, looks like I have some more work to do then: https://www.postgresql.org/docs/current/extend-type-system.html#EXTEND-TYPES-POLYMORPHIC

>Thus, when more than one argument position is declared with a polymorphic type, the net effect is that only certain combinations of actual argument types are allowed. For example, a function declared as equal(anyelement, anyelement) will take any two input values, so long as they are of the same data type.

>When the return value of a function is declared as a polymorphic type, there must be at least one argument position that is also polymorphic, and the actual data type(s) supplied for the polymorphic arguments determine the actual result type for that call. For example, if there were not already an array subscripting mechanism, one could define a function that implements subscripting as subscript(anyarray, integer) returns anyelement. This declaration constrains the actual first argument to be an array type, and allows the parser to infer the correct result type from the actual first argument's type. Another example is that a function declared as f(anyarray) returns anyenum will only accept arrays of enum types.

yuzefovich commented:
Maybe just relax the assertion in case we have polymorphic types so that it's not "internal error" with a stack trace?

DrewKimball commented:
>Maybe just relax the assertion in case we have polymorphic types so that it's not "internal error" with a stack trace?

I think we'll end up using those polymorphic types for the function return type, which is definitely wrong. I think implementing the same restrictions as PG will probably solve it for us.
